src/handlers/testing/align/components/App.tsx

```diff
@@ -0,0 +1,218 @@
+    // autogenerating evaluators
+    isLoading: false,
+    data: undefined,
+  });
+
+  const isLastTestCaseHash = (hash: string) => {
+    const testCaseHashes = props.testCaseHashes();
+    return testCaseHashes[testCaseHashes.length - 1] === hash;
+  };
+
+  const isGradingLastTestCase: boolean =
+    currentTestCaseResult.data !== undefined &&
+    isLastTestCaseHash(currentTestCaseResult.data.testCaseHash);
+
+  const requestNextTestCaseResult = async (args: {
+    prevHash: string | undefined;
+  }) => {
+    const testCaseHashes = props.testCaseHashes();
+    const firstHash = testCaseHashes[0];
+
+    if (!firstHash) {
+      throw new Error('No test cases');
+    }
+
+    setCurrentTestCaseResult({
+      isLoading: true,
+      data: undefined,
+    });
+
+    if (args.prevHash) {
+      const prevHashIdx = testCaseHashes.indexOf(args.prevHash);
+      if (prevHashIdx === -1) {
+        throw new Error('Previous hash not found');
+      }
+      const nextTestCaseHash = testCaseHashes[prevHashIdx + 1];
+      if (!nextTestCaseHash) {
+        throw new Error('No more test cases');
+      }
+      await props.requestNextTestCaseResult(nextTestCaseHash);
+    } else {
+      // Request the first test case if no previous hash is provided
+      await props.requestNextTestCaseResult(firstHash);
+    }
+  };
+
+  const onTestCaseResultGraded = async (args: {
+    accepted: boolean;
+    reason: string | undefined;
+  }) => {
+    if (!currentTestCaseResult.data) {
+      return;
+    }
+
+    const gradedResult: GradedTestCaseResult = {
+      testExternalId: props.testExternalId,
+      testCaseHash: currentTestCaseResult.data.testCaseHash,
+      testCaseBody: currentTestCaseResult.data.testCaseBody,
+      testCaseOutput: currentTestCaseResult.data.testCaseOutput,
+      grade: {
+        type: 'binary',
+        accepted: args.accepted,
+        reason: args.reason,
+      },
+    };
+
+    setGradedTestCaseResults([...gradedTestCaseResults, gradedResult]);
+    await props.onTestCaseResultGraded(gradedResult);
+  };
+
+  const onAfterGradeAction = async (action: AfterGradeAction) => {
+    if (!currentTestCaseResult.data) {
+      return;
+    }
+
+    switch (action) {
+      case AfterGradeAction.GRADE_NEXT:
+        const prevHash = currentTestCaseResult.data.testCaseHash;
+        await requestNextTestCaseResult({ prevHash });
+        break;
+      case AfterGradeAction.STOP_GRADING:
+        setCurrentTestCaseResult({
+          isLoading: false,
+          data: undefined,
+        });
+        setAutogeneratedEvaluators({
+          isLoading: true,
+          data: undefined,
+        });
+        const autogeneratedEvaluators = await props.autogenerateEvaluators();
+        setAutogeneratedEvaluators({
+          isLoading: false,
+          data: autogeneratedEvaluators,
+        });
+        props.onExit();
+        inkApp.exit();
+        break;
+    }
+  };
+
+  useEffect(() => {
```

Comment:
Can you add a comment about intention of this useEffect

Reply:
i can add comments for creating / removing the event listener, is that what you mean?